storage-proofs/src/vde.rs

```diff
@@ -117,11 +128,20 @@ fn create_key<H: Hasher>(
     parents: &[usize],
     data: &[u8],
     m: usize,
+    ciphertexts: Option<&mut Vec<u8>>,
 ) -> Result<H::Domain> {
     // ciphertexts will become a buffer of the layout
     // id | encodedParentNode1 | encodedParentNode1 | ...
 
-    let mut ciphertexts = vec![0u8; 32 * (parents.len() + 1)];
+    let mut owned_vector: Vec<u8>;
+    let ciphertexts: &mut Vec<u8> = match ciphertexts {
+        Some(preallocated) => preallocated,
+        None => {
+            owned_vector = vec![0u8; 32 * (parents.len() + 1)];
+            owned_vector.as_mut()
```

Comment:
Please tell me there's a cleaner way to do this.

Reply:
this is already pretty good, though I would use `&mut [u8]` instead of `Vec`. One nicer alternative would be to reuse this scratch memory across calls using `thread_local`. You can see an example of this here: 

- https://github.com/dignifiedquire/num-bigint/blob/div/src/algorithms/div.rs#L147-L149
- https://github.com/dignifiedquire/num-bigint/blob/div/src/algorithms/div.rs#L217